File: backend/app/schemas/recurring.py
from pydantic import BaseModel, Field
from datetime import datetime, date
from decimal import Decimal
from typing import Optional


class RecurringCreate(BaseModel):
    payee_id: Optional[str] = None
    amount: Decimal
    category_id: Optional[str] = None
    frequency: str  # weekly, biweekly, monthly, quarterly, yearly
    next_date: date
    account_id: Optional[str] = None
    is_subscription: bool = False


class RecurringUpdate(BaseModel):
    payee_id: Optional[str] = None
    amount: Optional[Decimal] = None
    category_id: Optional[str] = None
    frequency: Optional[str] = None
    next_date: Optional[date] = None
    account_id: Optional[str] = None
    is_subscription: Optional[bool] = None


class RecurringSuggestionResponse(BaseModel):
    dedupe_key: str
    payee_id: str
    payee_name: str
    suggested_amount: float
    suggested_frequency: str
    occurrence_count: int
    last_date: date
    suggested_next_date: date
    confidence: float
    category_id: Optional[str] = None
    account_id: Optional[str] = None


class RecurringSuggestionDismissBody(BaseModel):
    dedupe_key: str = Field(..., min_length=1, max_length=128)


class PriceChangeResponse(BaseModel):
    """A subscription whose latest charge stepped up from its established price."""

    payee_name: str
    previous_amount: float
    current_amount: float
    pct_change: float


class RecurringResponse(BaseModel):
    id: str
    household_id: str
    payee_id: Optional[str]
    payee_name: Optional[str] = None
    amount: Decimal
    category_id: Optional[str]
    category_name: Optional[str] = None
    frequency: str
    next_date: date
    account_id: Optional[str]
    account_name: Optional[str] = None
    is_subscription: bool
    created_at: datetime

    model_config = {"from_attributes": True}

File: frontend/src/lib/api/recurring.ts
import api from "./client";

export interface RecurringTransaction {
  id: string;
  household_id: string;
  payee_id: string | null;
  payee_name: string | null;
  amount: number;
  category_id: string | null;
  category_name: string | null;
  frequency: string;
  next_date: string;
  account_id: string | null;
  account_name: string | null;
  is_subscription: boolean;
  created_at: string;
}

export interface RecurringCreate {
  payee_id?: string;
  amount: number;
  category_id?: string;
  frequency: string;
  next_date: string;
  account_id?: string;
  is_subscription?: boolean;
}

export interface RecurringSuggestion {
  dedupe_key: string;
  payee_id: string;
  payee_name: string;
  suggested_amount: number;
  suggested_frequency: string;
  occurrence_count: number;
  last_date: string;
  suggested_next_date: string;
  confidence: number;
  category_id: string | null;
  account_id: string | null;
}

export interface SubscriptionPriceChange {
  payee_name: string;
  previous_amount: number;
  current_amount: number;
  pct_change: number;
}

export const recurringApi = {
  list: () => api.get<RecurringTransaction[]>("/recurring").then((r) => r.data),
  create: (data: RecurringCreate) =>
    api.post<RecurringTransaction>("/recurring", data).then((r) => r.data),
  update: (id: string, data: Partial<RecurringCreate>) =>
    api.put<RecurringTransaction>(`/recurring/${id}`, data).then((r) => r.data),
  delete: (id: string) => api.delete(`/recurring/${id}`),
  suggestions: (lookback_days = 90) =>
    api
      .get<RecurringSuggestion[]>("/recurring/suggestions", { params: { lookback_days } })
      .then((r) => r.data),
  dismissSuggestion: (dedupe_key: string) =>
    api.post("/recurring/suggestions/dismiss", { dedupe_key }),
  priceChanges: () =>
    api.get<SubscriptionPriceChange[]>("/recurring/price-changes").then((r) => r.data),
};
